### ISSUE TITLE
Can't open any .ipynb or .py file

### ISSUE THREAD

**Issue by whichLain** [rationalmatter/Juno-Issue-Tracker#418]
**Describe the bug**
Can not open any notebook or .py file. I guess it's a safari problem, but can't make it sure.

The following description may provide some clues:

**April to July, 2024.**
Around April, I discovered that when using some VPNs, I couldn't use Jupyter Lab in Safari. Specifically, I could **open the page**, but I couldn't interact with the terminal. At the same time, if I used the same VPN, Juno Connect behaved the same as it does now, i.e., unable to open notebook files.
Until I updated to iOS 18 beta, I was unable to open notebook files when running this VPN software.
The same phenomenon could be observed in some late versions of iPadOS 17 and macOS 14. Using "Inspect" in macOS Safari, I found some "ws://xxxxx" elements reporting errors in the web page. However, the same VPN would not affect the Chrome browser.

**July to August, 2024.**
**Regardless of whether a VPN is used:** Juno Connect cannot open notebook files(iOS18 / iPadOS18 / macOS15).
**Without using a VPN:** Safari can normally use Jupyter Lab.
**Using a VPN:** On iOS 18 and iPadOS 18, Safari **cannot open Jupyter Lab pages**; on macOS Sequoia, Safari can normally use Jupyter Lab.

**To Reproduce**
1. Go to 'Jupyter Servers'
2. Tap on 'my self-hosted server'
3. Scroll down to 'any notebook file'

**Expected behavior**
I hope Juno team can help determine if this is indeed a Safari issue.
Considering the core infrastructure status of the Safari engine in iOS/iPadOS, if this issue is not fixed, similar software and functions like Juno Connect will be unusable.

**Screenshots**
![IMG_2352](https://github.com/user-attachments/assets/ff2826c0-45b0-4cfd-a6bb-4768fabe4bf7)
![IMG_2353](https://github.com/user-attachments/assets/205f02f6-c0f9-495d-ae59-872e9d012de8)
![IMG_2355](https://github.com/user-attachments/assets/ac8811b2-0169-4a2a-9537-f477b0c3d870)

**Environment (please complete the following information):**
 - Device: [iPhone 15 Pro]
 - iOS version: [iOS 18.0 public beta]
 - App version [1.6.13]
 - Local Network is allowed in settings.

Thank you very much!

**Comment by alexstaravoitau** [comment]
Hi @whichLain — I suspect you may be connecting to a server running JupyterLab >4.0 or Jupyter Notebook >7.0 (see #365), in which case you may need to install `nbclassic` and update your connection settings (see [this comment](https://github.com/rationalmatter/Juno-Issue-Tracker/issues/365#issuecomment-1676341107)). Basically, the comment suggests the following:

1. Install "classic" Notebook web app with `pip install nbclassic`.
2. Update the server's URL in Juno Connect to `{SERVER_ADDRESS}/nbclassic/tree`.

I'm still looking into supporting JupyterLab 4.0 in a more seamless manner (apologies for inconvenience meanwhile 😔).

Now, let me know if you still experience the issues with connecting the web sockets, because depending on your server configuration this could be a SSL certificate issue (I think).

**Comment by sebaro1906** [comment]
I‘m having the same issue. And I think it is something with the SSL certificate. I use tljh with 

sudo tljh-config set https.enabled true
sudo tljh-config set https.letsencrypt.email 

Any ideas?

**Comment by alexstaravoitau** [comment]
@sebaro1906 Please check if you're connecting to JupyterLab 4.0+ / Jupyter Notebook 7.0+, this would be the most likely cause of the issue (see steps above on how to work around that).

**Comment by sebaro1906** [comment]
Yes, I already added „/nbclassic/tree“ to the adress, but that didn’t fix it.

**Comment by sebaro1906** [comment]
Also, it says „Not trusted“ in the connection setting. But the notebook should load anyway?

**Comment by alexstaravoitau** [comment]
@sebaro1906 I'm assuming you managed to install nbclassic first, right? Could you please attach screenshots of what you see when you try to connect to the server? Is this an SSH tunnel or a direct connection server?

Generally, servers marked as "Not trusted" will let you connect to the Jupyter server and open notebooks, but the kernels won't connect (or this was the behaviour several iOS versions ago, at least).

**Comment by sebaro1906** [comment]
@alexstaravoitau yes, I installed nbclassic first. It is a direct connection.  Attached you see a Screenshot of the connection. Does that help?

<img width="1170" height="2391" alt="Image" src="https://github.com/user-attachments/assets/58f907c0-229a-4073-9d20-b5ff8fec3b36" />

**Comment by sebaro1906** [comment]
This is the basic „tree“ view: 

<img width="1170" height="2406" alt="Image" src="https://github.com/user-attachments/assets/71cf1930-fe5d-435b-ab6b-33e194997297" />

**Comment by alexstaravoitau** [comment]
Thanks @sebaro1906 (and apologies for the late reply)—looks like it's the SSL cert blocking the kernel websocket. Two quick questions:

1. What URL did you enter in Juno Connect—the hostname Let's Encrypt was issued for, or an IP / .local name? A hostname mismatch alone makes the cert read as invalid.
2. Does Safari on the same device show that URL with a clean lock icon? If Safari complains too, the cert really is broken—Let's Encrypt likely never finished setting up on your server.

Either way, the most reliable fix is to trust the cert on the device manually as per walkthrough here: https://juno.sh/ssl-self-signed-cert/. I suppose only the _Install on iOS_ and _Enable Trust_ sections apply (skip the cert-generation steps at the top; export the cert your server already serves instead).

If this box is LAN-only, you can also just disable HTTPS and use plain HTTP—fine on a trusted network, not over the public internet.
